src/lib.rs

```diff
@@ -14,28 +14,51 @@ extern crate serde;
 #[macro_use]
 extern crate serde_derive;
 
+#[cfg(feature = "with-serde")]
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+
 use std::fmt;
 use std::net::IpAddr;
 
+mod common;
 mod ipv4;
 mod ipv6;
-mod common;
 
 use std::str::FromStr;
 
-pub use ipv4::{Ipv4Network, ipv4_mask_to_prefix};
-pub use ipv6::{Ipv6Network, ipv6_mask_to_prefix};
 pub use common::IpNetworkError;
+pub use ipv4::{ipv4_mask_to_prefix, Ipv4Network};
+pub use ipv6::{ipv6_mask_to_prefix, Ipv6Network};
 
 /// Represents a generic network range. This type can have two variants:
 /// the v4 and the v6 case.
-#[cfg_attr(feature = "with-serde", derive(Serialize, Deserialize))]
```

Comment:
Since `IpNetwork` is just an enum with tuple variants you don't need to implement manually here. The default derive will work if you "untag" it. That is:-

```rust
#[cfg_attr(feature = "with-serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "with-serde", serde(untagged))]
```